### ISSUE TITLE
expose intermediate outputs in map calculation

### ISSUE THREAD

**Issue by johnarevalo** [cytomining/copairs#36]
By disaggregating the `run_pipeline` function in a modular way

**Comment by shntnu** [comment]
@johnarevalo Does this issue also include the following feature?

I want to be able to know the number of positive and negative examples per row (which in turn is used to calculate the null)

**Comment by johnarevalo** [comment]
Yes. I think the values you are looking for are in `null_confs` variable inside `run_pipeline`:

https://github.com/cytomining/copairs/blob/491ebafbd8bb09088b9fdfd261b5f2b2157f0d93/src/copairs/map.py#L133

As quick workaround, you can copy the `run_pipeline` code in a separate script to inspect all the intermediate outputs.

**Comment by shntnu** [comment]
> As quick workaround, you can copy the `run_pipeline` code in a separate script to inspect all the intermediate outputs.

Thanks!

For our notes, here's what I did: https://github.com/broadinstitute/neuronal-cell-painting/blob/81bef05320e93c33f18f5587863845be63593697/ncp/src/my_run_pipeline.py

**Comment by shntnu** [comment]
Update: I figured there's no downside to doing this:

https://github.com/broadinstitute/neuronal-cell-painting/blob/2486dfa6402154192d4dd127bb000bc1f9755f7a/ncp/src/my_run_pipeline.py#L65-L68

**Comment by johnarevalo** [comment]
@shntnu , I forgot to mention  `null_confs` is a list of `[num_positive, total]`, not `[num_positive, num_negative]`.

PD: I just merged a typo Alan fixed. now it is `compute_ap_contiguous`.

**Comment by johnarevalo** [comment]
`n_pos_pairs` and `n_total_pairs` are now available in `v0.3.1`.

https://github.com/cytomining/copairs/blob/11570eb5e03d908e344bb8f7f90940e6dac603e3/src/copairs/map.py#L159-L160

**Comment by johnarevalo** [comment]
As of 0.4, we can have the pairs, the AP values, and the mAP values separately by invoking `matching.Matcher` (`matching.get_all_pairs` in 0.5), `compute.average_precision` and `compute.mean_average_precision`.

Num pos and num total are also returned in the ap computation.
